cirq/protocols/measurement_key_protocol.py

```diff
@@ -67,28 +84,62 @@ def measurement_key(val: Any, default: Any = RaiseTypeErrorIfNotProvided):
             returned NotImplemented) and also no default value was specified.
+        ValueError: `val` has multiple measurement keys.
     """
-    getter = getattr(val, '_measurement_key_', None)
-    result = NotImplemented if getter is None else getter()
-    if result is not NotImplemented:
-        return result
+    result = measurement_keys(val)
+
+    if len(result) == 1:
+        return result[0]
+
+    if len(result) > 1:
+        raise ValueError(f'Got multiple measurement keys ({result!r}) '
+                         f'from {val!r}.')
 
     if default is not RaiseTypeErrorIfNotProvided:
         return default
 
-    if getter is None:
-        raise TypeError(
-            "object of type '{}' has no _measurement_key_ method.".format(
-                type(val)))
+    raise TypeError(f"object of type '{type(val)}' had no measurement keys.")
+
+
+def measurement_keys(val: Any, *,
+                     allow_decompose: bool = True) -> Tuple[str, ...]:
+    """Gets the measurement keys of measurements within the given value.
+
+    Args:
+        val: The value which has the measurement key..
+        allow_decompose: Defaults to True. When true, composite operations that
+            don't directly specify their measurement keys will be decomposed in
+            order to find measurement keys within the decomposed operations. If
+            not set, composite operations will appear to have no measurement
+            keys.
+
+    Returns:
+        The measurement keys of the value. If the value has no measurement,
+        the result is the empty tuple.
+    """
+    getter = getattr(val, '_measurement_keys_', None)
+    result = NotImplemented if getter is None else getter()
+    if result is not NotImplemented and result is not None:
+        return tuple(result)
+
+    getter = getattr(val, '_measurement_key_', None)
+    result = NotImplemented if getter is None else getter()
+    if result is not NotImplemented and result is not None:
+        return result,
+
+    if allow_decompose:
+        operations, _, _ = _try_decompose_into_operations_and_qubits(val)
+        if operations is not None:
+            return tuple(
+                key for op in operations for key in measurement_keys(op))
 
-    raise TypeError("object of type '{}' does have a _measurement_key_ method, "
-                    "but it returned NotImplemented.".format(type(val)))
+    return ()
 
 
 def is_measurement(val: Any) -> bool:
-    """Returns whether or not the given value is a measurement.
+    """Determines whether or not the given value is a measurement.
 
-    A measurement must implement the `measurement_key` protocol and have a
-    channel, as represented by the `has_channel` protocol returning true.
+    Measurements are identified by the fact that `cirq.measurement_keys` returns
+    a non-empty result for them.
     """
-    return measurement_key(val, None) is not None and has_channel(val)
+    return bool(measurement_keys(val))
```

Comment:
Before we would consider objects that have a quantum channel to be measurements, probably to reflect the fact that non-trace preserving quantum operations are measurements.

I don't have the full story here, but I'm concerned we may be breaking some code (in cirq or elsewhere) that depends on this.

Reply:
The fallout of removing this check is that a bad operation that said it was a measurement but had no specified channel would fail slightly later. This is not really a useful check IMO. In particular, note that all the tests continued to pass.

Further note that some callers require the measurement to be something more specific than a channel (e.g. a stabilizer operation, an operation that can apply to a state vector, etc). Relying on this method to check that the measurement can act on your state in the way you need is coupling two things together that can be kept separate.